staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
 	StorageVersionHash string `json:"storageVersionHash,omitempty" protobuf:"bytes,10,opt,name=storageVersionHash"`
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Discovery API Types
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DiscoveryAPIGroupList struct {
+	TypeMeta `json:",inline"`
+	// groups is a list of APIGroup.
+	// +listType=map
+	// +listMapKey=name
+	Groups []DiscoveryAPIGroup `json:"groups" protobuf:"bytes,1,rep,name=groups"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type DiscoveryAPIGroup struct {
+	TypeMeta `json:",inline"`
+	// name is the name of the group.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// versions are the versions supported in this group.
+	// This will be sorted in descending order based on the preferred version
+	// +listType=map
+	// +listMapKey=version
+	Versions []DiscoveryGroupVersion `json:"versions" protobuf:"bytes,2,rep,name=versions"`
+}
+
+// GroupVersion contains the "group/version" and "version" string of a version.
+// It is made a struct to keep extensibility.
+type DiscoveryGroupVersion struct {
+	// version specifies the version in the form of "version". This is to save
+	// the clients the trouble of splitting the GroupVersion.
+	Version string `json:"version" protobuf:"bytes,1,opt,name=version"`
+	// resources contains the name of the resources and if they are namespaced.
+	// +listType=map
+	// +listMapKey=name
+	APIResources []DiscoveryAPIResource `json:"resources" protobuf:"bytes,2,rep,name=resources"`
+
+	// LastContacted is the last time that the apiserver has successfully reached the
+	// corresponding group version's discovery document. This will be nil if the group-version
+	// has not been aggregated yet (APIResources will be empty). To maintain consistency across scenarios with multiple
+	// apiservers, this time will be quantized down to the nearest fifteen minutes.
+	// LastContacted *time.Time `json:"lastContacted" protobuf:"bytes,3,opt,name=lastContacted"`
+}
+
+// APIResource specifies the name of a resource and whether it is namespaced.
+type DiscoveryAPIResource struct {
+	// name is the plural name of the resource.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// singularName is the singular name of the resource.  This allows clients to handle plural and singular opaquely.
+	// The singularName is more correct for reporting status on a single item and both singular and plural are allowed
+	// from the kubectl CLI interface.
+	SingularName string `json:"singularName" protobuf:"bytes,6,opt,name=singularName"`
+	// namespaced indicates if a resource is namespaced or not.
+	Namespaced bool `json:"namespaced" protobuf:"varint,2,opt,name=namespaced"`
```

Comment:
in keeping with modern standards, can this be an enum for Scope with value Cluster and Namespace?

Reply:
+1